# Unidentified Flying Object (UFO) Sightings - Initial Exploration

This notebook explores a dataset of 80,000+ reported UFO sightings around the world between November 11, 1906, and August 08, 2014.

## 1. Import Required Libraries

In [303]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

plt.style.use("default")

## 2. Load the Raw UFO Dataset

In [304]:
df = pd.read_csv("../data/ufo_sightings_raw.csv")

C:\Users\nlnz8\AppData\Local\Temp\ipykernel_124244\1935580793.py:1: DtypeWarning: Columns (5,9) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("../data/ufo_sightings_raw.csv")


## 3. Inspect & Understand the Dataset Structure

### 3.1 Preview the Raw Dataset

In [305]:
df.head(10)

datetime                  city state country     shape  \
0  1949-10-10 20:30:00            san marcos    tx      us  cylinder   
1  1949-10-10 21:00:00          lackland afb    tx     NaN     light   
2  1955-10-10 17:00:00  chester (uk/england)   NaN      gb    circle   
3  1956-10-10 21:00:00                  edna    tx      us    circle   
4  1960-10-10 20:00:00               kaneohe    hi      us     light   
5  1961-10-10 19:00:00               bristol    tn      us    sphere   
6  1965-10-10 21:00:00    penarth (uk/wales)   NaN      gb    circle   
7  1965-10-10 23:45:00               norwalk    ct      us      disk   
8  1966-10-10 20:00:00             pell city    al      us      disk   
9  1966-10-10 21:00:00              live oak    fl      us      disk   

  duration (seconds) duration (hours/min)  \
0               2700           45 minutes   
1               7200              1-2 hrs   
2                 20           20 seconds   
3                 20             1/2 hour   
4                900           15 minutes   
5                300            5 minutes   
6                180         about 3 mins   
7               1200           20 minutes   
8                180           3  minutes   
9                120      several minutes   

                                            comments date posted    latitude  \
0  This event took place in early fall around 194...  2004-04-27  29.8830556   
1  1949 Lackland AFB&#44 TX.  Lights racing acros...  2005-12-16    29.38421   
2  Green/Orange circular disc over Chester&#44 En...  2008-01-21        53.2   
3  My older brother and twin sister were leaving ...  2004-01-17  28.9783333   
4  AS a Marine 1st Lt. flying an FJ4B fighter/att...  2004-01-22  21.4180556   
5  My father is now 89 my brother 52 the girl wit...  2007-04-27      36.595   
6  penarth uk  circle  3mins  stayed 30ft above m...  2006-02-14   51.434722   
7  A bright orange color changing to reddish colo...  1999-10-02     41.1175   
8  Strobe Lighted disk shape object observed clos...  2009-03-19  33.5861111   
9  Saucer zaps energy from powerline as my pregna...  2005-05-11  30.2947222   

   longitude   
0  -97.941111  
1  -98.581082  
2   -2.916667  
3  -96.645833  
4 -157.803611  
5  -82.188889  
6   -3.180000  
7  -73.408333  
8  -86.286111  
9  -82.984167

**Notes**
- Some values (e.g., NaN) in "state" and/or "country" are present.
- Text fields appear inconsistent in formatting (e.g., varied capitalization, mixed formatting, and spacing).

### 3.2 Inspect Column Types and Missing Values

In [306]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80332 entries, 0 to 80331
Data columns (total 11 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   datetime              80332 non-null  object 
 1   city                  80332 non-null  object 
 2   state                 74535 non-null  object 
 3   country               70662 non-null  object 
 4   shape                 78400 non-null  object 
 5   duration (seconds)    80332 non-null  object 
 6   duration (hours/min)  80332 non-null  object 
 7   comments              80317 non-null  object 
 8   date posted           80332 non-null  object 
 9   latitude              80332 non-null  object 
 10  longitude             80332 non-null  float64
dtypes: float64(1), object(10)
memory usage: 6.7+ MB


**Notes**:
- Several columns that should contain datetime or numeric values (e.g., "latitude", "duration" (seconds), "datetime", "date posted") were initially loaded as "object". 
- Erroneous dtypes prevent time-series, numeric, and geospatial analysis. 
- Several missing values present in multiple columns, especially "country", "state", and "comments"

### 3.3 Summary Statistics for All Columns

In [307]:
df.describe(include="all")

datetime     city  state country  shape duration (seconds)  \
count                 80332    80332  74535   70662  78400              80332   
unique                69474    19900     67       5     29                705   
top     2010-07-04 22:00:00  seattle     ca      us  light                300   
freq                     36      525   9655   65114  16565               7070   
mean                    NaN      NaN    NaN     NaN    NaN                NaN   
std                     NaN      NaN    NaN     NaN    NaN                NaN   
min                     NaN      NaN    NaN     NaN    NaN                NaN   
25%                     NaN      NaN    NaN     NaN    NaN                NaN   
50%                     NaN      NaN    NaN     NaN    NaN                NaN   
75%                     NaN      NaN    NaN     NaN    NaN                NaN   
max                     NaN      NaN    NaN     NaN    NaN                NaN   

       duration (hours/min)  comments date posted    latitude    longitude   
count                 80332     80317       80332       80332  80332.000000  
unique                 8304     79997         317       23292           NaN  
top               5 minutes  Fireball  2009-12-12  47.6063889           NaN  
freq                   4716        11        1510         481           NaN  
mean                    NaN       NaN         NaN         NaN    -86.772885  
std                     NaN       NaN         NaN         NaN     39.697205  
min                     NaN       NaN         NaN         NaN   -176.658056  
25%                     NaN       NaN         NaN         NaN   -112.073333  
50%                     NaN       NaN         NaN         NaN    -87.903611  
75%                     NaN       NaN         NaN         NaN    -78.755000  
max                     NaN       NaN         NaN         NaN    178.441900

**Notes**:
- High cardinality in text-based columns (e.g., 19,900 unique cities). *High cardinality = harder to group/visualize, harder to encode, slower/more expensive ops.*
- Many stat fields are NaN because the corresponding columns are stored as objects instead of numeric/datetime.
- Most common city ("seattle"), country ("us"), and shape ("light").

### 3.4 Dataset Dimensions

In [209]:
df.shape

(80332, 11)

**Notes**: 
The dataset contains 11 columns, and 80,332 reported UFO sightings, capturing time, location, duration, and descriptive comments. 

### 3.5 Explore Key Categorical Columns

#### 3.5.1 UFO Shapes Reported

In [309]:
df["shape"].value_counts()

shape
light        16565
triangle      7865
circle        7608
fireball      6208
other         5649
unknown       5584
sphere        5387
disk          5213
oval          3733
formation     2457
cigar         2057
changing      1962
flash         1328
rectangle     1297
cylinder      1283
diamond       1178
chevron        952
egg            759
teardrop       750
cone           316
cross          233
delta            7
round            2
crescent         2
pyramid          1
flare            1
hexagon          1
dome             1
changed          1
Name: count, dtype: int64

**Notes**:
- It appears that the three most sighted shapes between 1906 and 2014 were "light", "triangle", and "circle". 
- The "changed" column is vague. Does "changed" indicate the witness saw the object change shape, or is it a data entry artifact? This may require further investigation. 
- The "other" category is quite significant (5649). The significant number of reports in the "other" category suggests ambiguous or inconsistent reporting. What types of descriptions fall into this label? Could this category be refined? 

#### 3.5.2 Country with the Most Sightings

In [211]:
df["country"].value_counts()

country
us    65114
ca     3000
gb     1905
au      538
de      105
Name: count, dtype: int64

**Notes**:
- The country with the most reported UFO sightings is the United States (U.S.). 
- Why are sightings so heavily concentrated in the U.S.? 
- Does this reflect true frequency, reporting behavior, population size, or data collection bias? 
- To what extent might variables such as a larger population equipped with advanced technology (for instance, internet access and recording devices), or increased interest from organizations like MUFON and NUFORC, contribute to observational or reporting biases regarding UFO phenomena?

#### 3.5.3 U.S. States with the Most Sightings

In [310]:
df["state"].value_counts().head(10)

state
ca    9655
wa    4268
fl    4200
tx    3677
ny    3219
az    2689
il    2645
pa    2582
oh    2425
mi    2071
Name: count, dtype: int64

**Notes**:
- According to our data, UFO reported sightings in the U.S. are most likely to occur in California, Washington, Florida, and Texas.
- Comparing sightings across different states, California overwhelmingly stands out. 
- Is California's high count due to population size, visibility conditions, reporting culture, or dataset bias? 
- Certain UFO shapes (e.g., "light", "triangle", and "circle") are reported more often than others.

#### 3.5.4 Most Frequent Datetime Entries

In [315]:
df["datetime"].value_counts().head(10)

datetime
2010-07-04 22:00:00    36
2012-07-04 22:00:00    31
1999-11-16 19:00:00    27
2009-09-19 20:00:00    26
2011-07-04 22:00:00    25
2004-10-31 20:00:00    23
2010-07-04 21:00:00    23
2013-07-04 22:00:00    22
2012-07-04 22:30:00    21
1999-11-16 19:05:00    20
Name: count, dtype: int64

**Notes**:
- Why do certain timestamps appear frequently? 
- What was happening historically during these times/dates? 
- Further analyses may concentrate on the years with the highest frequency of reported sightings, allowing for closer examination of these subsets and comparative analysis across different years.

### 3.6 Missing Values Check

The following section represents early data quality checks; full cleaning will occur in later phases. 

In [214]:
# Identifying Missing Values
df.isna().sum()

datetime                   0
city                       0
state                   5797
country                 9670
shape                   1932
duration (seconds)         0
duration (hours/min)       0
comments                  15
date posted                0
latitude                   0
longitude                  0
dtype: int64

**Notes**: 
- Several columns contain missing values, especially those related to location fields (e.g., state & country). 

### 3.7 Duplicate Records Check

In [317]:
df.duplicated().sum()

np.int64(0)

## 4. Initial Observations and Guiding Questions:

**A. Data Observations:**
- most columns are categorical ("object"), type conversion will be necessary for time-based and numerical analysis.
- missing values need to be addressed, specifically ("State" and "Country") to obtain accurate geographic comparisons.
- "Duration" data will need to be standardized due to inconsistent formatting before analysis.
- Reported UFO sightings seemed to be heavily concentrated in the U.S., especially California. This needs to be noted and investigated to determine if reporting bias has occurred.  

**B. Guiding Questions:**

The following questions focus on temporal, geographic, and descriptive patterns:

1. How have reported UFO sightings changed over time (by year and decade), and are there observable global trends?
2. Are increases in sightings consistent across countries, or are they primarily driven by reports from the United States?
3. Are certain UFO shapes more commonly reported in specific geographic regions?
4. Do reported UFO shapes vary across different time periods or decades?
5. What temporal patterns exist in UFO sightings (e.g., time of day, seasonality)?
6. Are certain UFO shapes associated with longer or shorter reported sighting durations?
7. Does the concentration of sightings in the U.S. appear stable over time, or does it fluctuate across decades?
8. To what extent might reporting frequency be influenced by population density, access to technology, or cultural factors, as inferred through geographic patterns?

**C. Extended & Future Questions Beyond the Current Dataset:**

The following questions fall outside the scope of the current dataset and are noted as potential long-term extensions requiring additional external data sources:

1. Are reported UFO sightings more common near military bases or government facilities?
2. Is there a geographic relationship between reported UFO sightings and reported cattle mutilation incidents?
3. Given that “UFO” is a broad classification, are there external datasets that could help refine categories (e.g., UAPs "Unidentified Aerial Phenomenon", USOs "Unidentified Submerged Object", interdimensional, or other unidentified phenomena) beyond simple shape descriptions?
4. Can additional datasets help contextualize sightings using environmental, military, or atmospheric data to move beyond descriptive classifications?

## 5. Data Cleaning & Preparation for EDA

### 5.1 Hidden Missing Values Detection
Objective:
Identify non-standard missing-value placeholders to safely replace them. 

In [216]:
# Define all hidden missing-value patterns: 

hidden_missing = [
    "", " ", "  ", "   ", 
    "?", "??", "-", "--", "---",
    "unknown", "Unknown",
    "n/a", "N/A", "na", "NA",
    "none", "None",
    "null", "Null",
    ".", "..", "...",
    "nan", "NaN"
]

# Count hidden missing values in each column before replacement
hidden_counts = df.apply(lambda col: col.astype(str).isin(hidden_missing).sum())
hidden_counts

datetime                   0
city                      43
state                   5797
country                 9670
shape                   7516
duration (seconds)         0
duration (hours/min)       0
comments                  16
date posted                0
latitude                   0
longitude                  0
dtype: int64

**Notes**
- The raw dataset contains many non-standard missing-value placeholders such as "?", "na", "none", and text variations like "Unknown".
- These must be identified and replaced to avoid incorrect type conversion or misleading summary statistics. 

### 5.2 Standardizing Missing Values to NaN

Objective: Convert all invalid placeholder entries to proper NaN, enabling consistent downstream processing. 

In [217]:
# Replace hidden missing values with np.nan
df = df.replace(hidden_missing, np.nan)

**Notes**:
- Replacing invalid entries with NaN ensures pandas correctly detects missing data (i.e., converting columns to datetime or numeric formats in later phases).

### 5.3 Verify Missing Value Fixes

In [218]:
# Recalcuate missing values after replacement 
missing_after_cleanup = df.isna().sum()

missing_after_cleanup

datetime                   0
city                      43
state                   5797
country                 9670
shape                   7516
duration (seconds)         0
duration (hours/min)       0
comments                  16
date posted                0
latitude                   0
longitude                  0
dtype: int64

**Notes**:
- Some columns (e.g. "state" and "country") still contain legitimate missing values after replacing placeholders.
- True missing values will be addressed in the geographic cleanup step.

### 5.4 Check Missing Value Percentages

This summary shows the percentage of missing values in each column after replacing hidden palceholders with NaN. 

In [219]:
# Percentage of missing values per column 
missing_pct = (df.isna().sum() / len(df) * 100).round(2)
missing_pct

datetime                 0.00
city                     0.05
state                    7.22
country                 12.04
shape                    9.36
duration (seconds)       0.00
duration (hours/min)     0.00
comments                 0.02
date posted              0.00
latitude                 0.00
longitude                0.00
dtype: float64

**Notes**: "country" (12.04%), "shape" (9.36%), and "state" (7.22%) represent the highest % of missing values.

### 5.5 Geographic Cleanup

Objective: Before using latitude/longitude to infer missing locations, clean the fields so only true missing values remain.

*Note: "unknown" = term is used as a temporary placeholder, proper geospatial inference will replace these later.*

#### 5.5a Inspect Missing Geographic Values

In [220]:
# Check missing geographic fields
geo_missing = df[["state", "country"]].isna().sum()
geo_missing

state      5797
country    9670
dtype: int64

**Notes**: 
- The high count of missing UFO-sighting location data for states (5,797) and country (9,670) indicates that the geographic info is incomplete in my records.

- The numerous missing "state" and "country" values confirm that many geographic fields were not originally standardized. 
- These missing values will require inference later (via coordinates and shapefile-based spatial joins).

#### 5.5b Standardize Geographic Fields

Purpose: Ensure "state" and "country" values are consistently formatted (lowercase, no whitespace) so comparisons and merging efficiency.

In [221]:
# Standardize"state" and "country" fields
for col in ["state", "country"]:
    df[col] = (
        df[col]
        .astype("string")
        .str.strip()
        .str.lower()
        .replace("", pd.NA)
        .replace("?", pd.NA)
    )

**Note:** 
- "state" and "country" converted to a consistent text format (i.e., lowercase, trimmed).
- Replaces empty or invalid strings so missing values are accurately recognized. 

#### 5.5c Fill Temporary "unknown"

In [222]:
df["state"] = df["state"].fillna("unknown")
df["country"] = df["country"].fillna("unknown")

**Note:** 
- "unknown" is used as a temporary placeholder so that missing values can be tracked.
- Spatial joins in Section 7 will replace unknown entries with accurate geographic assignments/lookup.

#### 5.5d Verify Fixes 

In [223]:
# Check Missing Datetime Count Again
df[["state", "country"]].isna().sum()

state      0
country    0
dtype: int64

**Note:** Confirmed no blank or invalid geographic values remain after standardization.

## 6. Import Geopandas & Load Shapefiles

### 6.1 Import GeoPandas 

In [224]:
import geopandas as gpd

**Notes**:
- GeoPandas is required for geospatial and mapping. 

### 6.2 Load World & U.S. Shapefiles

In [225]:
world_path = "../data/shapefiles/world/ne_50m_admin_0_countries.shp"
gdf_world = gpd.read_file(world_path)

states_path = "../data/shapefiles/us_states/cb_2022_us_state_5m.shp"
gdf_states = gpd.read_file(states_path)

**Notes**: The shapefiles corresponding to world layers ("world_path") and state layers ("state_path") supply the necessary polygon data required for accurately mapping UFO sightings by country and by U.S. state.

### 6.3 Preview the Data

In [226]:
gdf_world.head(), gdf_states.head()

(        featurecla  scalerank  LABELRANK SOVEREIGNT SOV_A3  ADM0_DIF  LEVEL  \
 0  Admin-0 country          1          3   Zimbabwe    ZWE         0      2   
 1  Admin-0 country          1          3     Zambia    ZMB         0      2   
 2  Admin-0 country          1          3      Yemen    YEM         0      2   
 3  Admin-0 country          3          2    Vietnam    VNM         0      2   
 4  Admin-0 country          5          3  Venezuela    VEN         0      2   
 
                 TYPE TLC      ADMIN  ... FCLASS_TR  FCLASS_ID FCLASS_PL  \
 0  Sovereign country   1   Zimbabwe  ...      None       None      None   
 1  Sovereign country   1     Zambia  ...      None       None      None   
 2  Sovereign country   1      Yemen  ...      None       None      None   
 3  Sovereign country   1    Vietnam  ...      None       None      None   
 4  Sovereign country   1  Venezuela  ...      None       None      None   
 
   FCLASS_GR  FCLASS_IT FCLASS_NL FCLASS_SE  FCLASS_BD FCLAS

In [227]:
gdf_states.columns

Index(['STATEFP', 'STATENS', 'AFFGEOID', 'GEOID', 'STUSPS', 'NAME', 'LSAD',
       'ALAND', 'AWATER', 'geometry'],
      dtype='object')

In [228]:
gdf_states.head()

STATEFP   STATENS     AFFGEOID GEOID STUSPS         NAME LSAD         ALAND  \
0      35  00897535  0400000US35    35     NM   New Mexico   00  314198573403   
1      72  01779808  0400000US72    72     PR  Puerto Rico   00    8869029522   
2      48  01779801  0400000US48    48     TX        Texas   00  676685555821   
3      21  01779786  0400000US21    21     KY     Kentucky   00  102266581101   
4      39  01085497  0400000US39    39     OH         Ohio   00  105823621267   

        AWATER                                           geometry  
0    726463825  POLYGON ((-109.05017 31.48, -109.04984 31.4995...  
1   4922249087  MULTIPOLYGON (((-65.3357 18.34954, -65.32933 1...  
2  18974391187  POLYGON ((-106.64548 31.89867, -106.64084 31.9...  
3   2384240769  MULTIPOLYGON (((-89.40565 36.52816, -89.39868 ...  
4  10274734976  MULTIPOLYGON (((-82.73571 41.60336, -82.72309 ...

**Notes**: .head() confirms that the geometry fields loaded correctly and the files were read successfully. 

### 6.4 Check CRS (Coordinate Reference Systems)

In [318]:
gdf_world.crs, gdf_states.crs

(<Geographic 2D CRS: EPSG:4326>
 Name: WGS 84
 Axis Info [ellipsoidal]:
 - Lat[north]: Geodetic latitude (degree)
 - Lon[east]: Geodetic longitude (degree)
 Area of Use:
 - name: World.
 - bounds: (-180.0, -90.0, 180.0, 90.0)
 Datum: World Geodetic System 1984 ensemble
 - Ellipsoid: WGS 84
 - Prime Meridian: Greenwich,
 <Geographic 2D CRS: EPSG:4326>
 Name: WGS 84
 Axis Info [ellipsoidal]:
 - Lat[north]: Geodetic latitude (degree)
 - Lon[east]: Geodetic longitude (degree)
 Area of Use:
 - name: World.
 - bounds: (-180.0, -90.0, 180.0, 90.0)
 Datum: World Geodetic System 1984 ensemble
 - Ellipsoid: WGS 84
 - Prime Meridian: Greenwich)

**Notes**: 
-	U.S. Census shapefiles use NAD83, but NAD83 is not the same as WGS 84. Cannot special-join layers with different CRS. 
-	Shapefiles must share the same CRS before spatial joins or plotting. 
-	The world layer uses EPSG:4326, but the U.S. states layer uses EPSG:4269, requiring re-projection.

### 6.5 Reproject U.S. States to Match World CRS (EPSG:4326)

In [319]:
gdf_states = gdf_states.to_crs("EPSG:4326")

# Verify change
gdf_states.crs

<Geographic 2D CRS: EPSG:4326>
Name: WGS 84
Axis Info [ellipsoidal]:
- Lat[north]: Geodetic latitude (degree)
- Lon[east]: Geodetic longitude (degree)
Area of Use:
- name: World.
- bounds: (-180.0, -90.0, 180.0, 90.0)
Datum: World Geodetic System 1984 ensemble
- Ellipsoid: WGS 84
- Prime Meridian: Greenwich

**Notes**: 
- The U.S. states layer is now aligned with global coordinates.
- After reprojection, both world and U.S. layers share EPSG:4326, enabling accurate spatial operations.

## 7. Convert UFO Coordinates Into Geographic Information

### 7.1 Clean Column Names 

Purpose: Some column names contain trailing spaces, which break merge operations and comparisons. Stripping whitespace ensures consistent behavior. 

In [231]:
# Remove leading/trailing spaces in column names
df.columns = df.columns.str.strip()

df.columns

Index(['datetime', 'city', 'state', 'country', 'shape', 'duration (seconds)',
       'duration (hours/min)', 'comments', 'date posted', 'latitude',
       'longitude'],
      dtype='object')

**Note**: clean column names to prevent merge failures caused by trailing spaces.

### 7.2 Import Point/Geo Libraries

In [232]:
# Import Point and geopandas
from shapely.geometry import Point
import geopandas as gpd

**Note**: GeoPandas and Shapely are required to convert latitude/longitude into point geometries. 

### 7.3 Clean Latitude & Longitude

Purpose: Convert text coordinates to numeric values and detect invalid entries. 

In [233]:
# Convert latitude & longitude to numeric (fix invalid values)
df["latitude"] = pd.to_numeric(df["latitude"], errors="coerce")
df["longitude"] = pd.to_numeric(df["longitude"], errors="coerce")

# Check how many coordinates are missing
df[["latitude", "longitude"]].isna().sum()

latitude     1
longitude    0
dtype: int64

**Notes**:
- Only one invalid latitude value was converted to NaN and manually corrected..
- Missing (NaN) or invalid values were converted into NaN during pd.to_numeric(..., errors="coerce")
- All coordinates now numeric -> required for mapping and spatial joins.

In [234]:
# Show specific column values with missing latitude
df[df["latitude"].isna()][["datetime", "city", "state", "country", "comments", "latitude", "longitude"]]

datetime                          city state  country  \
43782  1974-05-22 05:30:00  mescalero indian reservation    nm  unknown   

                                                comments  latitude   longitude  
43782  Huge rectangular object emmitting intense whit...       NaN -105.624152

In [320]:
# Mescalero Indian Reservation latitude:
df.loc[43782, "latitude"] = 33.33
df.loc[43782, "country"] = "us"

In [236]:
# Verification 
df.loc[43782][["city", "state", "country", "latitude", "longitude"]]

city         mescalero indian reservation
state                                  nm
country                                us
latitude                            33.33
longitude                     -105.624152
Name: 43782, dtype: object

In [237]:
# Check how many coordinates are missing
df[["latitude", "longitude"]].isna().sum()

latitude     0
longitude    0
dtype: int64

**Note**:
I decided not to drop the row as I was able to approximate the latitude within a relatively small margin of error as it was located within a specific indian resservation in New Mexico, and the longitude was also provided.

### 7.4 Convert UFO Data Into GeoDataFrame

Purpose: Add a geometry column so the dataset can be used in spatial joins, mapping, and geospatial analysis.

In [238]:
# Create a copy to avoid modifying original df
gdf_ufo = df.copy()

# Create geometry column
gdf_ufo["geometry"] = gdf_ufo.apply(
    lambda row: Point(row["longitude"], row["latitude"]), axis=1
)

# Convert to GeoDataFrame
gdf_ufo = gpd.GeoDataFrame(gdf_ufo, geometry="geometry", crs="EPSG:4326")

# Verification
gdf_ufo.head()

datetime                  city    state  country     shape  \
0  1949-10-10 20:30:00            san marcos       tx       us  cylinder   
1  1949-10-10 21:00:00          lackland afb       tx  unknown     light   
2  1955-10-10 17:00:00  chester (uk/england)  unknown       gb    circle   
3  1956-10-10 21:00:00                  edna       tx       us    circle   
4  1960-10-10 20:00:00               kaneohe       hi       us     light   

  duration (seconds) duration (hours/min)  \
0               2700           45 minutes   
1               7200              1-2 hrs   
2                 20           20 seconds   
3                 20             1/2 hour   
4                900           15 minutes   

                                            comments date posted   latitude  \
0  This event took place in early fall around 194...  2004-04-27  29.883056   
1  1949 Lackland AFB&#44 TX.  Lights racing acros...  2005-12-16  29.384210   
2  Green/Orange circular disc over Chester&#44 En...  2008-01-21  53.200000   
3  My older brother and twin sister were leaving ...  2004-01-17  28.978333   
4  AS a Marine 1st Lt. flying an FJ4B fighter/att...  2004-01-22  21.418056   

    longitude                     geometry  
0  -97.941111   POINT (-97.94111 29.88306)  
1  -98.581082   POINT (-98.58108 29.38421)  
2   -2.916667        POINT (-2.91667 53.2)  
3  -96.645833   POINT (-96.64583 28.97833)  
4 -157.803611  POINT (-157.80361 21.41806)

**Notes**:
- One invalid latitude value was converted to NaN and manually corrected.
- All coordinates now numeric -> required for mapping and spatial joins.

### 7.5 Spatial Join: Assign Countries Automatically

Purpose: Match each UFO sighting to a country boundary polygon.

In [239]:
# Keep only the world boundry geometry + country name
world = gdf_world[["ADMIN", "geometry"]].rename(columns={"ADMIN": "country_filled"})

# Spatial join to assign country by position
ufo_with_country = gpd.sjoin(
    gdf_ufo,
    world,
    how="left",
    predicate="within"
)

# Inspect results
ufo_with_country[["country", "country_filled"]].head()

country            country_filled
0       us  United States of America
1  unknown  United States of America
2       gb            United Kingdom
3       us  United States of America
4       us  United States of America

**Notes**:
- The spatial join assigns each UFO sighting to a country’s polygon based on its point coordinates.
- Using predicate="within" ensures a point is matched only if it lies inside a country's boundaries.
- World country polygons are used to fill out missing or incorrect country labels.

In [240]:
ufo_with_country.columns

Index(['datetime', 'city', 'state', 'country', 'shape', 'duration (seconds)',
       'duration (hours/min)', 'comments', 'date posted', 'latitude',
       'longitude', 'geometry', 'index_right', 'country_filled'],
      dtype='object')

In [241]:
# Remove leftover index_right column from previous spatial join
if "index_right" in ufo_with_country.columns:
    ufo_with_country = ufo_with_country.drop(columns=["index_right"])

### 7.6 Spatial Join for U.S. States

Purpose: Match each U.S. UFO sighting to a state boundry polygon, to render a clean "state_filled" column.

In [242]:
# Keep only state name + geometry for U.S. states
states = gdf_states[["NAME", "geometry"]].rename(columns={"NAME": "state_filled"})

# Spatial join to assign U.S. state by position
ufo_with_state = gpd.sjoin(
    ufo_with_country,
    states,
    how="left",
    predicate="within"
)

# Inspect results
ufo_with_state[["state", "state_filled"]].head()

state state_filled
0       tx        Texas
1       tx        Texas
2  unknown          NaN
3       tx        Texas
4       hi       Hawaii

**Note**: This spatial join resolves missing or incorrect U.S. state values by determining the state in which each UFO coordinate falls.


### 7.7 Clean & Standardize U.S. States Alphabetically

Purpose: Create a standardized mapping between full state names and their two-letter codes to ensure consistency across all future analyses.

In [243]:
# Build mapping automatically from shapefile in alphabetical order
state_name_to_code = (
    gdf_states.set_index("NAME")["STUSPS"]
    .str.lower()    # convert to lowercase for consistency
    .sort_index()  # sorts alphabetically by state name
    .to_dict()
)

state_name_to_code

{'Alabama': 'al',
 'Alaska': 'ak',
 'American Samoa': 'as',
 'Arizona': 'az',
 'Arkansas': 'ar',
 'California': 'ca',
 'Colorado': 'co',
 'Commonwealth of the Northern Mariana Islands': 'mp',
 'Connecticut': 'ct',
 'Delaware': 'de',
 'District of Columbia': 'dc',
 'Florida': 'fl',
 'Georgia': 'ga',
 'Guam': 'gu',
 'Hawaii': 'hi',
 'Idaho': 'id',
 'Illinois': 'il',
 'Indiana': 'in',
 'Iowa': 'ia',
 'Kansas': 'ks',
 'Kentucky': 'ky',
 'Louisiana': 'la',
 'Maine': 'me',
 'Maryland': 'md',
 'Massachusetts': 'ma',
 'Michigan': 'mi',
 'Minnesota': 'mn',
 'Mississippi': 'ms',
 'Missouri': 'mo',
 'Montana': 'mt',
 'Nebraska': 'ne',
 'Nevada': 'nv',
 'New Hampshire': 'nh',
 'New Jersey': 'nj',
 'New Mexico': 'nm',
 'New York': 'ny',
 'North Carolina': 'nc',
 'North Dakota': 'nd',
 'Ohio': 'oh',
 'Oklahoma': 'ok',
 'Oregon': 'or',
 'Pennsylvania': 'pa',
 'Puerto Rico': 'pr',
 'Rhode Island': 'ri',
 'South Carolina': 'sc',
 'South Dakota': 'sd',
 'Tennessee': 'tn',
 'Texas': 'tx',
 'United States

**Note**: Generated a dictionary to convert state names to standardized 2-letter USPS codes.

In [244]:
# Fix list-type states: convert ["texas"] --> "texas"
ufo_clean["state_final"] = ufo_clean["state_final"].apply(
    lambda x: x[0] if isinstance(x, list) and len(x) > 0 else x
)

# Verifty fix
ufo_clean["state_final"].apply(type).value_counts()

state_final
<class 'str'>    80332
Name: count, dtype: int64

### 7.8 Create a Cleaned & Standardized state column

Purpose: Combine original state values with shapefile-derived values to produce one final, reliable state column for analysis.

In [245]:
# Making a safe working copy of the UFO dataframe to avoid overwriting earlier versions
ufo_clean = ufo_with_state.copy()

# Lowercase both "state" and "state_filled" to ensure comparisons are consistent
ufo_clean["state"] = ufo_clean["state"].astype(str).str.lower()
ufo_clean["state_filled"] = ufo_clean["state_filled"].astype(str).str.lower()

# Build final state - Define function to resolve the final state value
def resolve_state(row):
    state = row["state"]
    filled = row["state_filled"]

    # If original state is missing or labeled as unknown
    if state in ["unknown", "", None] or pd.isna(state) or state == "nan":

        # If shapefile successfully assigned as state
        if filled not in ["nan", "none", "", None] and not pd.isna(filled):
            return filled
        else:
            return "unknown"   # Last resort if no info was found anywhere
    # Otherwise: keep original state (already standardized)
    return state

# Apply the state resolution function -> creates a clean final state column
ufo_clean["state_final"] = ufo_clean.apply(resolve_state, axis=1)

# Preview Results
ufo_clean[["state", "state_filled", "state_final"]].head()

state state_filled state_final
0       tx        texas          tx
1       tx        texas          tx
2  unknown          nan     unknown
3       tx        texas          tx
4       hi       hawaii          hi

**Note**: The final state_final column merges original and inferred values for best accuracy.

**Logic:**
- If orginal state is missing -> use shapefile-dervied state_filled
- If shapefile value is available -> prefer state_filled
- If everything fails -> return "unknown"


### 7.9 Validate All Unique State Values

In [246]:
ufo_clean.columns

Index(['datetime', 'city', 'state', 'country', 'shape', 'duration (seconds)',
       'duration (hours/min)', 'comments', 'date posted', 'latitude',
       'longitude', 'geometry', 'country_filled', 'index_right',
       'state_filled', 'state_final'],
      dtype='object')

In [247]:
# Review all unique state codes
sorted(ufo_clean["state_final"].dropna().unique())

['ab',
 'ak',
 'al',
 'alabama',
 'ar',
 'arizona',
 'arkansas',
 'az',
 'bc',
 'ca',
 'california',
 'co',
 'colorado',
 'connecticut',
 'ct',
 'dc',
 'de',
 'fl',
 'florida',
 'ga',
 'georgia',
 'guam',
 'hi',
 'ia',
 'id',
 'il',
 'in',
 'kansas',
 'kentucky',
 'ks',
 'ky',
 'la',
 'louisiana',
 'ma',
 'massachusetts',
 'mb',
 'md',
 'me',
 'mi',
 'michigan',
 'mississippi',
 'missouri',
 'mn',
 'mo',
 'montana',
 'ms',
 'mt',
 'nb',
 'nc',
 'nd',
 'ne',
 'nevada',
 'new jersey',
 'new mexico',
 'new york',
 'nf',
 'nh',
 'nj',
 'nm',
 'north carolina',
 'ns',
 'nt',
 'nv',
 'ny',
 'oh',
 'ohio',
 'ok',
 'on',
 'or',
 'oregon',
 'pa',
 'pe',
 'pennsylvania',
 'pq',
 'pr',
 'puerto rico',
 'qc',
 'ri',
 'sa',
 'sc',
 'sd',
 'sk',
 'south carolina',
 'texas',
 'tn',
 'tx',
 'united states virgin islands',
 'unknown',
 'ut',
 'va',
 'virginia',
 'vt',
 'wa',
 'washington',
 'west virginia',
 'wi',
 'wv',
 'wy',
 'wyoming',
 'yk',
 'yt']

**Note**: Validating unique values ensures the final state column contains only standardized and expected codes after the cleaning pipeline.

In [248]:
# Count occurances of each state
ufo_clean["state_final"].value_counts()

state_final
ca                9655
unknown           5685
wa                4268
fl                4200
tx                3677
                  ... 
nevada               1
guam                 1
new york             1
montana              1
north carolina       1
Name: count, Length: 101, dtype: int64

In [249]:
# Check for invalid 3+ letter state values
invalid_states = [s for s in ufo_clean["state_final"].unique() if isinstance(s, str) and len(s) > 2]
invalid_states

['unknown',
 'missouri',
 'alabama',
 'colorado',
 'texas',
 'connecticut',
 'arkansas',
 'kansas',
 'south carolina',
 'ohio',
 'puerto rico',
 'new mexico',
 'florida',
 'arizona',
 'united states virgin islands',
 'mississippi',
 'michigan',
 'wyoming',
 'oregon',
 'virginia',
 'kentucky',
 'california',
 'louisiana',
 'pennsylvania',
 'georgia',
 'washington',
 'nevada',
 'massachusetts',
 'new jersey',
 'west virginia',
 'guam',
 'new york',
 'montana',
 'north carolina']

**Note**: Confirmed that all states are valid USPS codes or "unknown".

### 7.10 Convert State Names into 2-Letter Codes

Purpose: Standardize all U.S. state values to consistent 2-letter codes so they can be merged, grouped, and mapped reliably.

In [250]:
# Convert to string
ufo_clean["state_final"] = ufo_clean["state_final"].astype(str)

# Convert full names --> abbreviations using dictionary
ufo_clean["state_final"] = ufo_clean["state_final"].map(
    lambda x: state_name_to_code.get(x.strip().title(), x.lower())
)

# Validate 
sorted(ufo_clean["state_final"].unique())

['ab',
 'ak',
 'al',
 'ar',
 'az',
 'bc',
 'ca',
 'co',
 'ct',
 'dc',
 'de',
 'fl',
 'ga',
 'gu',
 'hi',
 'ia',
 'id',
 'il',
 'in',
 'ks',
 'ky',
 'la',
 'ma',
 'mb',
 'md',
 'me',
 'mi',
 'mn',
 'mo',
 'ms',
 'mt',
 'nb',
 'nc',
 'nd',
 'ne',
 'nf',
 'nh',
 'nj',
 'nm',
 'ns',
 'nt',
 'nv',
 'ny',
 'oh',
 'ok',
 'on',
 'or',
 'pa',
 'pe',
 'pq',
 'pr',
 'qc',
 'ri',
 'sa',
 'sc',
 'sd',
 'sk',
 'tn',
 'tx',
 'unknown',
 'ut',
 'va',
 'vi',
 'vt',
 'wa',
 'wi',
 'wv',
 'wy',
 'yk',
 'yt']

**Note**: The sorted results confirm that all states have been converted into lowercase 2-letter codes with no remaining full names.

In [251]:
# Confirm data types
ufo_clean["state_final"].apply(type).value_counts()

state_final
<class 'str'>    80332
Name: count, dtype: int64

### 7.11 Create initial "country_final" column

Purpose: Combine original and shapefile-derived country values so that missing or unknown entries can be replaced with the correct geographic country.

In [252]:
# Create "country_final" using country_filled when original is unknown
ufo_clean["country_final"] = ufo_clean.apply(
    lambda row: row["country_filled"]
    if row["country"] in ["unknown", None, "", "nan"]
    else row["country"], 
    axis=1
)

# Convert all to lowercase for consistency
ufo_clean["country_final"] = ufo_clean["country_final"].str.lower()

# Preview results
ufo_clean[["country", "country_filled", "country_final"]].head()

country            country_filled             country_final
0       us  United States of America                        us
1  unknown  United States of America  united states of america
2       gb            United Kingdom                        gb
3       us  United States of America                        us
4       us  United States of America                        us

**Notes**: 
- All values are now lowercase and consistent. 
- Where the original country weas missing, the shapefile-assigned value is correctly used instead.

### 7.12 Convert full country names into 2-letter codes (dictionary method) 

Purpose: Convert the most common country names in the database into standard 2-letter ISO codes using a predefined dictionary.

In [253]:
# Dictionary to convert full country names to codes
country_name_to_code = {
    "united states of america": "us",
    "united kingdom": "gb",
    "canada": "ca",
    "australia": "au",
    "new zealand": "nz",
    "germany": "de",
    "france": "fr",
    "mexico": "mx"
}

In [254]:
# Apply conversion
ufo_clean["country_final"] = ufo_clean["country_final"].map(
    lambda x: country_name_to_code.get(x, x)    # keep same value if no match
)

# Preview results
ufo_clean[["country", "country_filled", "country_final"]].head()

country            country_filled country_final
0       us  United States of America            us
1  unknown  United States of America            us
2       gb            United Kingdom            gb
3       us  United States of America            us
4       us  United States of America            us

**Note**: Dictionary conversion successful standardized frequent country names; unusual or rare names remain unchanged for later cleanup.

### 7.13 Validate Country Values

Purpose: Review the variety of country values to identify inconsistent naming, misspellings, and non-standard geographic labels.

In [255]:
sorted(ufo_clean["country_final"].dropna().unique())

['afghanistan',
 'aland',
 'albania',
 'algeria',
 'antarctica',
 'argentina',
 'armenia',
 'aruba',
 'au',
 'austria',
 'azerbaijan',
 'bahrain',
 'bangladesh',
 'barbados',
 'belarus',
 'belgium',
 'belize',
 'bermuda',
 'bolivia',
 'bosnia and herzegovina',
 'botswana',
 'brazil',
 'british virgin islands',
 'brunei',
 'bulgaria',
 'ca',
 'cabo verde',
 'cambodia',
 'cameroon',
 'cayman islands',
 'chile',
 'china',
 'colombia',
 'costa rica',
 'croatia',
 'cuba',
 'curaçao',
 'cyprus',
 'czechia',
 'de',
 'democratic republic of the congo',
 'denmark',
 'dominican republic',
 'east timor',
 'ecuador',
 'egypt',
 'el salvador',
 'estonia',
 'eswatini',
 'ethiopia',
 'fiji',
 'finland',
 'fr',
 'french polynesia',
 'gb',
 'georgia',
 'ghana',
 'greece',
 'greenland',
 'guam',
 'guatemala',
 'guyana',
 'honduras',
 'hong kong s.a.r.',
 'hungary',
 'iceland',
 'india',
 'indonesia',
 'iran',
 'iraq',
 'ireland',
 'isle of man',
 'israel',
 'italy',
 'jamaica',
 'japan',
 'jersey',
 'jo

**Note**: The list shows many inconsistent entries, confirming the need for systematic standardization before mapping or aggregation.

In [256]:
# Ensure all country values are strings
ufo_clean["country_final"] = ufo_clean["country_final"].astype(str)

### 7.14 Convert Country Names to Standard 2-Letter Codes

Purpose: Use pycountry to automatically convert remaining country names into valid ISO 2-letter codes after dictionary cleanup.

In [257]:
import pycountry

def country_to_alpha2(country):
    if not isinstance(country, str):
        country = str(country)

    country = country.strip().lower()

    try:
        return pycountry.countries.lookup(country).alpha_2.lower()
    except:
        return country   # keep unchanged if no match

In [258]:
# Apply conversion
ufo_clean["country_final"] = ufo_clean["country_final"].apply(country_to_alpha2)

# Review results
sorted(ufo_clean["country_final"].dropna().unique())

['ae',
 'af',
 'al',
 'aland',
 'am',
 'aq',
 'ar',
 'at',
 'au',
 'aw',
 'az',
 'ba',
 'bb',
 'bd',
 'be',
 'bg',
 'bh',
 'bm',
 'bo',
 'br',
 'brunei',
 'bw',
 'by',
 'bz',
 'ca',
 'ch',
 'cl',
 'cm',
 'cn',
 'co',
 'cr',
 'cu',
 'cv',
 'cw',
 'cy',
 'cz',
 'de',
 'democratic republic of the congo',
 'dk',
 'do',
 'dz',
 'east timor',
 'ec',
 'ee',
 'eg',
 'es',
 'et',
 'fi',
 'fj',
 'fr',
 'gb',
 'ge',
 'gh',
 'gl',
 'gr',
 'gt',
 'gu',
 'gy',
 'hn',
 'hong kong s.a.r.',
 'hr',
 'hu',
 'id',
 'ie',
 'il',
 'im',
 'in',
 'iq',
 'ir',
 'is',
 'it',
 'je',
 'jm',
 'jo',
 'jp',
 'ke',
 'kg',
 'kh',
 'kosovo',
 'kr',
 'kw',
 'ky',
 'kz',
 'la',
 'lb',
 'lc',
 'lk',
 'ls',
 'lt',
 'lu',
 'lv',
 'ly',
 'ma',
 'mk',
 'mm',
 'mn',
 'mt',
 'mu',
 'mv',
 'mx',
 'my',
 'na',
 'nan',
 'ng',
 'nl',
 'no',
 'northern cyprus',
 'np',
 'nz',
 'om',
 'pa',
 'palestine',
 'pe',
 'pf',
 'ph',
 'pk',
 'pl',
 'pr',
 'pt',
 'py',
 'qa',
 'ro',
 'rs',
 'russia',
 'sa',
 'saint helena',
 'sb',
 'se',
 'sg',

**Note**: Some values may still convert incorrectly due to disputed territories or non-ISO regions; these will be fixed manually in the next step.

In [259]:
# Count how many rows per country
ufo_clean["country_final"].value_counts()

country_final
us                                  70682
ca                                   3587
gb                                   2352
au                                    630
nan                                   553
                                    ...  
ug                                      1
democratic republic of the congo        1
cw                                      1
gu                                      1
la                                      1
Name: count, Length: 158, dtype: int64

### 7.15 Fix incorrect country codes returned by pycountry

Purpose: Manually correct country values that pycountry cannot resolve or returns incorrectly, ensuring accurate and consistent ISO codes.

In [266]:
import numpy as np

# Manual correction dictionary for all erroneous or unrecognized values
# Keys MUST be lowercase with no surrounding spaces - since we normalize before lookup
country_corrections = {
    # Aaland / Åland Islands
    "aland": "ax",
    "åland": "ax",
    "aland islands": "ax",
    "åland islands": "ax",

    # Democratic Republic of Congo 
    "democratic republic of the congo": "cd",

    # East Timor / Timor-Leste 
    "east timor": "tl",
    "timor-leste": "tl",

    # Palestine 
    "palestine": "ps",
    "state of palestine": "ps",

    # Kosovo (pycountry does not support officially) 
    "kosovo": "xk",

    # Saint Helena 
    "saint helena": "sh",

    # Bahamas 
    "the bahamas": "bs",
    "bahamas": "bs",

    # São Tomé and Príncipe (multiple variations) 
    "são tomé and príncipe": "st",
    "sao tomé and príncipe": "st",
    "são tomé and principe": "st",
    "sao tome and principe": "st",
    "são tome and principe": "st",
    "sao tomé and principe": "st",

    # Brunei
    "brunei": "bn",

    # Russia 
    "russia": "ru",

    # Turkey 
    "turkey": "tr",

    # Hong Kong 
    "hong kong s.a.r.": "hk",
    "hong kong": "hk",

    # United States Virgin Islands 
    "united states virgin islands": "vi",
    "virgin islands": "vi",
    "usvi": "vi",

    # Values treated as missing 
    "northern cyprus": np.nan,
    "nan": np.nan,
    "none": np.nan,
    "": np.nan,
    " ": np.nan
}

# Apply corrections — normalizing case + whitespace first
def fix_country_name(x):
    if isinstance(x, str):
        x_clean = x.strip().lower()
        return country_corrections.get(x_clean, x_clean)
    return x

ufo_clean["country_final"] = ufo_clean["country_final"].apply(fix_country_name)

# Review cleaned unique list
sorted(ufo_clean["country_final"].dropna().unique())


['ae',
 'af',
 'al',
 'am',
 'aq',
 'ar',
 'at',
 'au',
 'aw',
 'ax',
 'az',
 'ba',
 'bb',
 'bd',
 'be',
 'bg',
 'bh',
 'bm',
 'bn',
 'bo',
 'br',
 'bs',
 'bw',
 'by',
 'bz',
 'ca',
 'cd',
 'ch',
 'cl',
 'cm',
 'cn',
 'co',
 'cr',
 'cu',
 'cv',
 'cw',
 'cy',
 'cz',
 'de',
 'dk',
 'do',
 'dz',
 'ec',
 'ee',
 'eg',
 'es',
 'et',
 'fi',
 'fj',
 'fr',
 'gb',
 'ge',
 'gh',
 'gl',
 'gr',
 'gt',
 'gu',
 'gy',
 'hk',
 'hn',
 'hr',
 'hu',
 'id',
 'ie',
 'il',
 'im',
 'in',
 'iq',
 'ir',
 'is',
 'it',
 'je',
 'jm',
 'jo',
 'jp',
 'ke',
 'kg',
 'kh',
 'kr',
 'kw',
 'ky',
 'kz',
 'la',
 'lb',
 'lc',
 'lk',
 'ls',
 'lt',
 'lu',
 'lv',
 'ly',
 'ma',
 'mk',
 'mm',
 'mn',
 'mt',
 'mu',
 'mv',
 'mx',
 'my',
 'na',
 'ng',
 'nl',
 'no',
 'np',
 'nz',
 'om',
 'pa',
 'pe',
 'pf',
 'ph',
 'pk',
 'pl',
 'pr',
 'ps',
 'pt',
 'py',
 'qa',
 'ro',
 'rs',
 'ru',
 'sa',
 'sb',
 'se',
 'sg',
 'sh',
 'si',
 'sk',
 'sn',
 'sr',
 'st',
 'sv',
 'sy',
 'sz',
 'tc',
 'th',
 'tl',
 'tn',
 'to',
 'tr',
 'tt',
 'tw',
 'ua',

**Note**: After applying corrections, remaining inconsistent or ambiguous values have been normalized or set to NaN where appropriate. 

### 7.16 Replace "unknown" and "nan" with NaN in country_final

Purpose: Standardize all palceholder country values ("unknown", "nan", empty strings) so they are treated as true missing values (NaN).

In [267]:
ufo_clean["country_final"] = ufo_clean["country_final"].replace(
    ["unknown", "nan", ""],
    np.nan
)

# Verify results
ufo_clean["country_final"].value_counts(dropna=False)

country_final
us     70682
ca      3587
gb      2352
au       630
NaN      554
       ...  
ug         1
cd         1
cw         1
gu         1
la         1
Name: count, Length: 157, dtype: int64

**Note**: The updated counts confirm that placeholder country labels were successfully removed, and only valid or missing ISO codes remain.

In [268]:
ufo_clean["country_final"].value_counts(dropna=False).head(20)

country_final
us     70682
ca      3587
gb      2352
au       630
NaN      554
mx       225
in       221
de       138
nl       112
za        91
nz        88
es        66
fr        65
br        63
my        52
jp        52
ie        51
be        40
no        38
pt        37
Name: count, dtype: int64

### 7.17 Final Country Validation Checklist

Purpose: Verify that rows missing "country_final" still contain valid latitude/longitude coordinates, ensuring they remain usable for later geographic assignment.

In [271]:
# Check if missing rows (NaN) have valid latitude/longitude
ufo_clean[ufo_clean["country_final"].isna()][["latitude", "longitude"]].head()

latitude   longitude
18   32.364167  -64.678611
77   40.935998  -73.901708
174  26.705621  -80.036430
366  21.344507 -157.974891
515 -33.137551   81.826172

**Notes**:
- Most rows with missing "country_final" values still contain valid geographic (lat/long) coordinates.
- Thes rows remain usable and can be accurately assigned to countries later if needed.

In [272]:
# Calculate the proportion of rows missing "country_final"
len(ufo_clean[ufo_clean["country_final"].isna()]) / len(ufo_clean)

0.006896380022904944

**Notes**:
- Only 0.0068963...(~0.69%) of all rows are missing "country_final".
- This is a small portion of the dataset and will not impact downstream geographic analysis. 

## 8 Build the Final Cleaned UFO Dataset 

### 8.1 Select Only Final Cleaned Columns

Purpose: Select only the finalized, fully cleaned columns needed for analysis and remove temporary or intermediate variables (e.g., "state_filled", "country_filled", "geometry", "index_right") to avoid redundancy and ensure clarity.

In [275]:
ufo_clean.columns

Index(['datetime', 'city', 'state', 'country', 'shape', 'duration (seconds)',
       'duration (hours/min)', 'comments', 'date posted', 'latitude',
       'longitude', 'geometry', 'country_filled', 'index_right',
       'state_filled', 'state_final', 'country_final'],
      dtype='object')

In [321]:
# Select only the cleaned, final columns for analysis
ufo_final = ufo_clean[[
    "datetime",
    "city",
    "state_final",
    "country_final",
    "shape",
    "duration (seconds)",
    "duration (hours/min)",
    "comments",
    "date posted",
    "latitude",
    "longitude"
]].copy()

# Preview final dataset
ufo_final.head(

)

datetime                  city state_final country_final  \
0  1949-10-10 20:30:00            san marcos          tx            us   
1  1949-10-10 21:00:00          lackland afb          tx            us   
2  1955-10-10 17:00:00  chester (uk/england)     unknown            gb   
3  1956-10-10 21:00:00                  edna          tx            us   
4  1960-10-10 20:00:00               kaneohe          hi            us   

      shape duration (seconds) duration (hours/min)  \
0  cylinder               2700           45 minutes   
1     light               7200              1-2 hrs   
2    circle                 20           20 seconds   
3    circle                 20             1/2 hour   
4     light                900           15 minutes   

                                            comments date posted   latitude  \
0  This event took place in early fall around 194...  2004-04-27  29.883056   
1  1949 Lackland AFB&#44 TX.  Lights racing acros...  2005-12-16  29.384210   
2  Green/Orange circular disc over Chester&#44 En...  2008-01-21  53.200000   
3  My older brother and twin sister were leaving ...  2004-01-17  28.978333   
4  AS a Marine 1st Lt. flying an FJ4B fighter/att...  2004-01-22  21.418056   

    longitude  
0  -97.941111  
1  -98.581082  
2   -2.916667  
3  -96.645833  
4 -157.803611

**Notes**:
- These columns form the cleaned UFO dataset used for all remaining analysis and EDA.
- All geographic fields ("state_final", "country_final") now use standardized 2-letter codes.
- Intermediate cleaning columns were intentionally excluded to maintain a clean, analysis-ready structure.

### 8.2 Sort the Final Cleaned UFO Dataset

Purpose: Sort the final dataset chronologically by the datetime column to ensure all temporal analyses reflect the true historical order of sightings.

In [279]:
# Sort final dataset by datetime (ascending, oldest -> newest)
ufo_final = ufo_final.sort_values(by="datetime").reset_index(drop=True)

# Preview sorted dataset
ufo_final.head()

datetime                           city state_final  \
0  1906-11-11 00:00:00                 wien (austria)     unknown   
1  1910-01-02 00:00:00              kirksville (near)          mo   
2  1910-06-01 15:00:00                    wills point          tx   
3  1916-04-05 13:00:00  france (above; from aircraft)     unknown   
4  1920-06-11 21:00:00                         cicero          in   

  country_final  shape duration (seconds) duration (hours/min)  \
0            at  other              10800                  3 h   
1            us   disk                120              minutes   
2            us  cigar                120            2 minutes   
3            fr  cigar                 60         about 1 min.   
4            us    NaN                 60             1 minute   

                                            comments date posted   latitude  \
0  The oldest professional photo of a UFO object ...  2002-12-23  48.208174   
1  Historical sighting (1903 - 1913) Northern Mis...  2005-09-15  40.194722   
2       Cigar shaped object moving from West to East  2005-04-16  32.709167   
3  ((NUFORC Note:  Possible hoax.  PD))  Saw 3 ci...  2004-03-09  46.227638   
4  ((NUFORC Note:  Probable hoax.  Note date.  PD...  2009-05-12  40.123889   

   longitude  
0  16.373819  
1 -92.583056  
2 -96.008056  
3   2.213749  
4 -86.013333

***Notes**
- Sorting prevents misleading trends when analyzing yearly or monthly patterns.
- Sorting ensures the first and last sightings will match the earliest and latest datetimes in Section 8.3

### 8.3 Validate the Time Range of the Dataset

Purpose: Confirm the earliest and latest UFO sightings in the cleaned dataset to verify that datetime parsing and sorting were performed correctly.

In [322]:
# Show the first (oldest) date in the dataset
ufo_final["datetime"].min()

'1906-11-11 00:00:00'

In [323]:
# Show the first row (earliest sighting recorded)
ufo_final.head(1)

datetime        city state_final country_final     shape  \
0  1949-10-10 20:30:00  san marcos          tx            us  cylinder   

  duration (seconds) duration (hours/min)  \
0               2700           45 minutes   

                                            comments date posted   latitude  \
0  This event took place in early fall around 194...  2004-04-27  29.883056   

   longitude  
0 -97.941111

In [324]:
# Show the most recent (latest) date in the dataset
ufo_final["datetime"].max()

'2014-05-08 18:45:00'

In [327]:
# Show the first row (earliest sighting recorded)
ufo_final.tail(1)

datetime    city state_final country_final  shape  \
80331  2013-09-09 23:00:00  edmond          ok            us  cigar   

      duration (seconds) duration (hours/min)  \
80331             1020.0           17 minutes   

                                                comments date posted  \
80331  2 witnesses 2  miles apart&#44 Red &amp; White...  2013-09-30   

        latitude  longitude  
80331  35.652778 -97.477778

**Note**: Validation confirms no corrupted or incorrectly parsed datetime entries remain.

### 8.4 Export the Final Dataset to CSV

Purpose: Save the cleaned UFO dataset to a CSV file for use in future notebooks, visualizations, and analysis steps.

In [328]:
# Export the cleaned, final dataset for future analysis
ufo_final.to_csv("ufo_final_cleaned.csv", index=False)

**Note**: The exported file contains only final, standardized fields ready for analysis.

### 8.5 Validate Saved File Loaded Correctly

Purpose: Confirm the exported dataset loads correctly and retains all expected columns and data types.

In [329]:
test = pd.read_csv("ufo_final_cleaned.csv")

test.info()
test.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80332 entries, 0 to 80331
Data columns (total 11 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   datetime              80332 non-null  object 
 1   city                  80289 non-null  object 
 2   state_final           80332 non-null  object 
 3   country_final         79778 non-null  object 
 4   shape                 72816 non-null  object 
 5   duration (seconds)    80332 non-null  object 
 6   duration (hours/min)  80332 non-null  object 
 7   comments              80316 non-null  object 
 8   date posted           80332 non-null  object 
 9   latitude              80332 non-null  float64
 10  longitude             80332 non-null  float64
dtypes: float64(2), object(9)
memory usage: 6.7+ MB


C:\Users\nlnz8\AppData\Local\Temp\ipykernel_124244\238246712.py:1: DtypeWarning: Columns (5) have mixed types. Specify dtype option on import or set low_memory=False.
  test = pd.read_csv("ufo_final_cleaned.csv")


datetime                  city state_final country_final  \
0  1949-10-10 20:30:00            san marcos          tx            us   
1  1949-10-10 21:00:00          lackland afb          tx            us   
2  1955-10-10 17:00:00  chester (uk/england)     unknown            gb   
3  1956-10-10 21:00:00                  edna          tx            us   
4  1960-10-10 20:00:00               kaneohe          hi            us   

      shape duration (seconds) duration (hours/min)  \
0  cylinder               2700           45 minutes   
1     light               7200              1-2 hrs   
2    circle                 20           20 seconds   
3    circle                 20             1/2 hour   
4     light                900           15 minutes   

                                            comments date posted   latitude  \
0  This event took place in early fall around 194...  2004-04-27  29.883056   
1  1949 Lackland AFB&#44 TX.  Lights racing acros...  2005-12-16  29.384210   
2  Green/Orange circular disc over Chester&#44 En...  2008-01-21  53.200000   
3  My older brother and twin sister were leaving ...  2004-01-17  28.978333   
4  AS a Marine 1st Lt. flying an FJ4B fighter/att...  2004-01-22  21.418056   

    longitude  
0  -97.941111  
1  -98.581082  
2   -2.916667  
3  -96.645833  
4 -157.803611